gordon/metrics/log.py

```diff
@@ -0,0 +1,173 @@
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Gordon ships with a default IMetricRelay implementation, which
+outputs metrics to the application logger using the standard
+library `logging` module.
+
+This implementation is used by default if no other IMetricRelay
+implementation is loaded and chosen.
+
+The default configuration that can be overriden is shown below.
+
+.. code-block:: ini
+
+    [metrics-logger]
+    # One of the log levels available from the stdlib logging module.
+    log_level = 'info'
+    # Time unit used by LogTimer. Can be one of 's', 'ms', 'ns'.
+    time_unit = 's'
+"""
+
+import collections
+import logging
+import time
+
+import zope.interface
+
+from gordon import interfaces
+
+# TODO: update plugin loader to load metrics logger as default
+
+
+@zope.interface.implementer(interfaces.ITimer)
+class LogTimer:
+    """Timer that outputs metrics to the logger.
+
+    Args:
+        metric (dict): Metric to output.
+        logger (.LoggerAdapter): Object used for outputting metrics.
+        time_unit (str): Unit for LogTimer output.
+            Options: s, ms, ns.
+    """
+    def __init__(self, metric, logger, time_unit):
+        self.metric = metric
+        self.logger = logger
+        self.time_multiplier = self._time_unit_to_multiplier(time_unit)
+
+    def _time_unit_to_multiplier(self, unit):
+        return {
+            's': 1,
+            'ms': 1E3,
+            'ns': 1E9
+        }[unit]
+
+    async def __aenter__(self):
+        await self.start()
+        return self
+
+    async def __aexit__(self, *args):
+        await self.stop()
+
+    async def start(self):
+        """Start the timer."""
+        self._start_time = time.perf_counter()
+
+    async def stop(self):
+        """Stop the timer and output the time delta to the logger."""
+        time_elapsed = (
+            time.perf_counter() - self._start_time) * self.time_multiplier
+        self.metric['value'] = time_elapsed
+        self.logger.log(self.metric)
+
+
+class LoggerAdapter:
+    """Configurable adapter that formats and outputs metrics.
+
+    Args:
+        level (str): Log level at which to output metric.
+    """
+    LOGFMT = '[{metric_name}] value: {value}'
+
+    def __init__(self, level):
+        self.level = level
+        self._logger = logging.getLogger('gordon.metrics-logger')
+
+    def log(self, metric):
+        """Format and output metric.
+
+        Args:
+            metric (dict): Complete metric.
+        """
+        message = self.LOGFMT.format(**metric)
+        if metric['context']:
+            message += ' context: {context}'.format(context=metric['context'])
```

Comment:
Would this look like:

```
[request-latency] value: 1 context: {'attr1': 'value1', 'attr2': 'value2'}
```

What do you think of expanding the context so it looks more readable/understandable:

```
[request-latency] value: 1 attr1: value1 attr2: value2
```

Reply:
Another option if more readable:
```
[request-latency] value=1 attr1=value1 attr2=value2
```

Reply:
Just talked about this IRL, but still unsure about how to proceed.

@descent3me @smstone @ocsw @RochesterinNYC any thoughts or suggestions about formatting metric attributes/context for stdout?

Reply:
I'm not too crazy about any of 'em, and I have no better suggestions. I'd even be okay with including `context: {}` just to avoid extra formatting+concatenation.  Basically I think it's good enough as is and as readable as any of 'em, so I can live with it, but I also won't complain if you want to change it.